cpu/lpc2387/periph/timer.c

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_lpc2387
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the low-level timer driver for the LPC2387
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include "periph_conf.h"
+#include "periph_cpu.h"
+#include "periph/timer.h"
+
+/**
+ * @brief   Check the board config to make sure we do not exceed max number of
+ *          timers
+ */
+#if TIMER_NUMF > 3
+#error "ERROR in timer configuration: too many timers defined"
+#endif
+
+/**
+ * @brief   UART device configurations
+ */
+static timer_isr_ctx_t isr_ctx[TIMER_NUMOF];
+
+/**
+ * @brief   Forward declarations for interrupt functions
+ * @{
+ */
+void tim_isr_0(void);
+void tim_isr_1(void);
+void tim_isr_2(void);
+void tim_isr_3(void);
+/** @} */
+
+/**
+ * @brief   Get the base pointer of a timer
+ */
+static inline lpc23xx_timer_t *get_dev(tim_t tim)
+{
+    switch (tim) {
+        case 0:
+            return TMR0;
+#if TIMER_NUMOF > 1
+        case 1:
+            return TMR1;
+#endif
+#if TIMER_NUMOF > 2
+        case 2:
+            return TMR2;
+#endif
+#if TIMER_NUMOF > 3
+        case 3:
+            return TMR3;
+#endif
+        default:
+            return NULL;
+    }
+}
+
+static inline void pwr_clk_and_isr(tim_t tim)
+{
+    switch (tim) {
+        case 0:
+            PCONP |= (1 << 1);
+            PCLKSEL0 &= ~(0x03 << 2);
+            PCLKSEL0 |=  (0x01 << 2);
+            install_irq(TIMER0_INT, &tim_isr_0, 1);
+            break;
+#if TIMER_NUMOF > 1
+        case 1:
+            PCONP |= (1 << 2);
+            PCLKSEL0 &= ~(0x03 << 4);
+            PCLKSEL0 |=  (0x01 << 4);
+            install_irq(TIMER1_INT, &tim_isr_1, 1);
+            break;
+#endif
+#if TIMER_NUMOF > 2
+        case 2:
+            PCONP |= (1 << 22);
+            PCLKSEL1 &= ~(0x03 << 12);
+            PCLKSEL1 |=  (0x01 << 12);
+            install_irq(TIMER2_INT, &tim_isr_2, 1);
+            break;
+#endif
+#if TIMER_NUMOF > 3
+        case 3:
+            PCONP |= (1 << 23);
+            PCLKSEL1 &= ~(0x03 << 14);
+            PCLKSEL1 |=  (0x01 << 14);
+            install_irq(TIMER3_INT, &tim_isr_3, 1);
+            break;
+#endif
+    }
+}
+
+int timer_init(tim_t tim, unsigned int us_per_tick, void (*callback)(int))
+{
+    /* get the timers base register */
+    lpc23xx_timer_t *dev = get_dev(tim);
+
+    /* make sure the timer device is valid */
+    if (dev == NULL) {
+        return -1;
+    }
+
+    /* save the callback */
+    isr_ctx[tim].cb = callback;
+    /* enable power, config periph clock and install ISR vector */
```

Comment:
I do prefer empty lines before comments, but I case that's a style question that @OlegHahm hasn't inscribed in any rule, yet.